server/server.go

```diff
@@ -0,0 +1,825 @@
+}
+
+func (c *client) GetJob() (job interface{}, pool *worker.Pool, ok bool) {
+	err := c.conn.SetReadDeadline(time.Now().Add(c.timeout))
+	if err != nil {
+		// TODO: Is it possible for the client closing this half of the connection
+		// to cause SetReadDeadline to return io.EOF? If so, we may want to do the
+		// same logic as the other error handling block in this function.
+		c.Log(err)
+		c.conn.Close()
+		return nil, nil, false
+	}
+
+	pkt := new(protocol.Packet)
+	_, err = pkt.ReadFrom(c.conn)
+	if err != nil {
+		if err == io.EOF {
+			// We can't rule out the possibility that the client just closed the
+			// writing half of their connection (the reading half of ours), but still
+			// wants to receive responses. Thus, we don't kill the connection.
+			//
+			// We also don't call c.Log because the writer goroutine could, in the
+			// future, encounter an error that we legitimately want logged. Even if no
+			// "real" error is encountered, when the other half of the connection is
+			// closed, the writer goroutine will encounter EOF, and will log it, so
+			// even if the connection is closed correctly, it will still get logged.
+			log.Debugf("connection %v: reading half closed by client", c.name)
+		} else {
+			c.Log(err)
+			c.conn.Close()
+		}
+		return nil, nil, false
+	}
+
+	switch pkt.Operation.Opcode {
+	case protocol.OpECDSASignMD5SHA1, protocol.OpECDSASignSHA1,
+		protocol.OpECDSASignSHA224, protocol.OpECDSASignSHA256,
+		protocol.OpECDSASignSHA384, protocol.OpECDSASignSHA512:
+		return pkt, c.ecdsaPool, true
+	default:
+		return pkt, c.otherPool, true
+	}
+}
+
+func (c *client) SubmitResult(result interface{}) (ok bool) {
+	resp := result.(response)
+	pkt := protocol.Packet{
+		Header: protocol.Header{
+			MajorVers: 0x01,
+			MinorVers: 0x00,
+			Length:    resp.op.Bytes(),
+			ID:        resp.id,
+		},
+		Operation: resp.op,
+	}
+
+	buf, err := pkt.MarshalBinary()
+	if err != nil {
+		// According to MarshalBinary's documentation, it will never return a
+		// non-nil error.
+		panic(fmt.Sprintf("unexpected internal error: %v", err))
+	}
+
+	_, err = c.conn.Write(buf)
+	if err != nil {
+		c.Log(err)
+		c.conn.Close()
+		return false
+	}
+	return true
+}
+
+func (c *client) IsAlive() bool {
+	var buf [0]byte
+	_, err := c.conn.Read(buf[:])
+	return err == nil
+}
+
+func (c *client) Destroy() {
+	c.Log(nil)
+	c.conn.Close()
+}
+
+func (c *client) Log(err error) {
+	// Use a sync.Once so that only the first goroutine to encounter an error gets
+	// to log it. This avoids the circumstance where a goroutine encounters an
+	// error, logs it, and then closes the network connection, which causes the
+	// other goroutine to also encounter an error (due to the closed connection)
+	// and spuriously log it.
+	//
+	// We also use this to allow Destroy to block the reader or writer from
+	// logging anything at all by calling Log(nil).
+	c.err.Do(func() {
```

Comment:
To be fair, if there were a way to check whether a given error was due to an operation being performed on a closed connection, we could just use that, and I'll admit it'd be a lot cleaner and simpler. Unfortunately, I don't think anything like that exists (I've looked and haven't found anything).